--- config.sub.old	2022-02-13 14:56:49.268884174 -0500
+++ config.sub	2022-02-13 14:57:18.905551866 -0500
@@ -1376,7 +1376,7 @@
 	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
-	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es*)
+	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* | -atlas*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
